feat(*): polar co-ordinates, de moivre, trig identities, quotient group for angles

abhimanyupallavisudhir opened the pull request:
Several additions (polar form of a complex number with proof of uniqueness etc, De Moivre's theorem, sum-to-product and other trig identities, quotient group for angles) to `exponential.lean`, plus some dependencies in algebra.group_power

abhimanyupallavisudhir committed:
feat(*): nth roots, de moivre, trig identities

notes: nth roots had to be placed in analysis.exponential to avoid circular imports

jlpaca committed:
feat(*): equal_angle is an equivalence relation

abhimanyupallavisudhir committed:
chore(*) made theorems easier to use

abhimanyupallavisudhir committed:
Merge branch 'xena' of https://github.com/leanprover-community/mathlib into xena

abhimanyupallavisudhir committed:
chore (*) remove unnecessary parameters that were being annoying

abhimanyupallavisudhir committed:
chore(*) Moved group power theorems to group_power file

abhimanyupallavisudhir committed:
chore(*) added attribution

jlpaca committed:
chore(*) add attribution

kckennylau committed:
chore(*): clean up

ChrisHughes24 commented on algebra/group_power.lean:
```diff
@@ -471,6 +471,22 @@ theorem pow_nonneg {a : α} (H : 0 ≤ a) : ∀ (n : ℕ), 0 ≤ a ^ n
+  apply mul_lt_mul ih (le_of_lt Hxy) Hxpos (le_of_lt (pow_pos (lt_trans Hxpos Hxy) _))
+end
+
+theorem pow_eq {x y : α} {n : ℕ} (Hxpos : 0 < x) (Hypos : 0 < y) (Hnpos : 0 < n) (Hxyn : x ^ n = y ^ n) : x = y :=
```
`pow_right_inj` is a better name here.

sgouezel replied:
This should hold under the weaker assumption that x and y are nonnegative, right?

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,22 @@ by simp [log, lt_irrefl]
 exp_injective $ by rw [exp_log zero_lt_one, exp_zero]
 
+noncomputable definition nth_root (x : ℝ) (n : ℕ) : ℝ := exp ((log x) / n)
+
```
`nth_root` of `0` should be `zero` I think.

kbuzzard replied:
even if n=0?

jcommelin replied:
Sure, the only reasonable solution to `x^0 = 0` would be `x = 0`, right? For all other `x` it is clear that `x^0 = 1`.

digama0 replied:
Don't make me argue for 0^0 = 1 again...

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,22 @@ by simp [log, lt_irrefl]
+  (Hxpos : 0 < x) (Hypos : 0 < y) (Hnpos : 0 < n) (Hynx : y ^ n = x) :
+  y = nth_root x n :=
+pow_eq Hypos nth_root_pos Hnpos (Hynx.trans (nth_root_power Hxpos Hnpos).symm)
+
```
rather than having `y ^ n = x` as a hypothesis, just prove `nth_root (y ^. n) n = y`. Also `x`, `y` and `n` should be implicit.

abhimanyupallavisudhir replied:
@ChrisHughes24 Ok, I've made the changes.

abhimanyupallavisudhir committed:
chore(*): cleanup

abhimanyupallavisudhir committed:
fix (analysis/exponential): forgot to remove sorry

(theorems already present in algebra/group_power) -- should compile correctly now.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,36 @@ by simp [log, lt_irrefl]
+
+theorem nth_root_nonneg {x n} : nth_root x n ≥ 0 := 
+begin
+  cases classical.em(x = 0),
```
You can also write `by_cases h : x = 0,` for this.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,36 @@ by simp [log, lt_irrefl]
+
+theorem nth_root_unique {x y : ℝ} {n : ℕ}
+(Hxpos : 0 < x) (Hypos : 0 < y) (Hnpos : 0 < n) (Hynx : y ^ n = x) : y = nth_root x n 
+:= pow_right_inj Hypos (nth_root_pos (ne_of_gt Hxpos)) Hnpos (Hynx.trans (nth_root_power Hxpos Hnpos).symm)
```
Subsequent lines of a `theorem` statement should be indented, and `:=` should come at the end of a line.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,36 @@ by simp [log, lt_irrefl]
+:= pow_right_inj Hypos (nth_root_pos (ne_of_gt Hxpos)) Hnpos (Hynx.trans (nth_root_power Hxpos Hnpos).symm)
+
+theorem nth_root_unique' {x : ℝ} {n : ℕ} (Hxpos : 0 < x) (Hnpos : 0 < n) : nth_root (x ^ n) n = x 
+:= eq.symm (nth_root_unique (pow_pos Hxpos n) Hxpos Hnpos rfl)
```
This `:=` should also move to the previous line.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,36 @@ by simp [log, lt_irrefl]
 exp_injective $ by rw [exp_log zero_lt_one, exp_zero]
 
+noncomputable definition nth_root (x : ℝ) (n : ℕ) : ℝ := 
+if x = 0 then 0 else exp ((log x) / n)
```
It may be nice to have a lemma like
```lean
lemma nth_root_def {x : R} (hx : x \ne 0) {n : N} : nth_root x n = exp (log x / n) :=
dif_neg hx -- or whatever the proof really is
```
to make it easier to manipulate `nth_root` directly without relying on `simp`.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -397,6 +453,72 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+@[simp] lemma coe_add (x y : ℝ) : ↑(x + y : ℝ) = (↑x + ↑y : angle) := rfl
+@[simp] lemma coe_neg (x : ℝ) : ↑(-x : ℝ) = -(↑x : angle) := rfl
+@[simp] lemma coe_sub (x y : ℝ) : ↑(x - y : ℝ) = (↑x - ↑y : angle) := rfl
+@[simp] lemma coe_gsmul (x : ℝ) (n : ℤ) : ↑(gsmul n x : ℝ) = gsmul n (↑x : angle) :=
```
Hmm, does this not follow by some lemma from `angle.is_add_group_hom`?

ChrisHughes24 replied:
yes `is_add_group_hom.gsmul`. Also, these lemmas should be in the `angle` namespace, which itself should be in the `real` namespace, which I think it already is.

rwbarton commented on analysis/exponential.lean:
```diff
@@ -397,6 +453,72 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+  exact h.symm
+end
+
+theorem cos_sin_inj (θ ψ : ℝ) (Hcos : real.cos θ = cos ψ) (Hsin : real.sin θ = sin ψ) : equal_angle θ ψ :=
```
Are these `real.`s necessary here?

rwbarton commented on analysis/exponential.lean:
```diff
@@ -127,6 +127,36 @@ by simp [log, lt_irrefl]
+noncomputable definition nth_root (x : ℝ) (n : ℕ) : ℝ := 
+if x = 0 then 0 else exp ((log x) / n)
+
+theorem exp_mul (x : ℝ) : ∀ n : ℕ, exp(n*x) = (exp(x))^n
```
This lemma seems out of place here.

abhimanyupallavisudhir replied:
@rwbarton  All done.

rwbarton reviewed (commented):
This looks very nice!

abhimanyupallavisudhir committed:
chore (*): add nth_root_def, move exp_mul, other minor fixes

abhimanyupallavisudhir committed:
fix(*): Removed nth_root stuff

ChrisHughes24 commented on algebra/group_power.lean:
```diff
@@ -471,6 +471,22 @@ theorem pow_nonneg {a : α} (H : 0 ≤ a) : ∀ (n : ℕ), 0 ≤ a ^ n
+  { refl },
+  { exact absurd Hxyn (ne_of_gt (pow_lt hyx Hypos Hnpos)) },
+end
+
```
`0 \le x` and `0 \le y` is sufficient here.

ChrisHughes24 commented on data/complex/exponential.lean:
```diff
@@ -409,6 +409,10 @@ by rw lim_mul_lim;
+lemma exp_mul (x : ℝ) : ∀ n : ℕ, exp(n*x) = (exp(x))^n
+| 0 := by rw [nat.cast_zero, zero_mul, exp_zero, pow_zero]
+| (nat.succ n) := by rw [pow_succ', nat.cast_add_one, add_mul, exp_add, ←exp_mul, one_mul]
+
```
This should probably be called `exp_nat_mul`

ChrisHughes24 replied:
since this is in the `complex` namespace, `x` should be a complex.

ChrisHughes24 commented on data/complex/exponential.lean:
```diff
@@ -673,6 +685,10 @@ by simp [real.exp]
+lemma exp_mul (x : ℝ) : ∀ n : ℕ, exp(n*x) = (exp(x))^n
+| 0 := by rw [nat.cast_zero, zero_mul, exp_zero, pow_zero]
+| (nat.succ n) := by rw [pow_succ', nat.cast_add_one, add_mul, exp_add, ←exp_mul, one_mul]
+
```
This should probably be called `exp_nat_mul`

abhimanyupallavisudhir replied:
I'll edit that in -- Do you know why the build isn't compiling? It works on my computer. I'll upgrade my mathlib and check.

I'll PR the nth_root stuff as real_power as Mario suggested.

ChrisHughes24 replied:
You should merge the latest version of mathlib

abhimanyupallavisudhir committed:
fix(data\complex\exponential.lean): conflict with other changes

Commit:
Merge remote-tracking branch 'leanprover/master' into HEAD

Commit:
fix build

abhimanyupallavisudhir committed:
Update exponential.lean

abhimanyupallavisudhir committed:
Update exponential.lean

abhimanyupallavisudhir committed:
Update exponential.lean

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -397,6 +423,72 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+definition equal_angle (θ ψ : ℝ) := (θ : angle) = ψ
+definition semiequal_angle_even (θ ψ : ℝ) := (θ + ψ : angle) = 0
+definition semiequal_angle_odd (θ ψ : ℝ) := (θ + ψ : angle) = π
+
```
I would get rid of these definitions and just state the same theorems using the unfolded definitions. Also `θ = - ψ` is probably a slightly more useful conclusion of a theorem than `θ + ψ = 0`

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -397,6 +423,72 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+end
+
+theorem cos_sin_inj (θ ψ : ℝ) (Hcos : cos θ = cos ψ) (Hsin : sin θ = sin ψ) : equal_angle θ ψ :=
+begin
```
Make `θ` and `ψ` implicit here

ChrisHughes24 commented on algebra/group_power.lean:
```diff
@@ -509,6 +509,22 @@ theorem pow_nonneg {a : α} (H : 0 ≤ a) : ∀ (n : ℕ), 0 ≤ a ^ n
+  rw [pow_add, pow_add, nat.succ_eq_add_one, pow_one, pow_one],
+  apply mul_lt_mul ih (le_of_lt Hxy) Hxpos (le_of_lt (pow_pos (lt_trans Hxpos Hxy) _))
+end
+
```
should be called `pow_lt_pow_of_lt_left`

sgouezel replied:
Here again this should hold assuming only that x is nonnegative. There are maybe other instances in this PR where your assumptions can be weakened from positive to nonnegative, please check this.

abhimanyupallavisudhir committed:
fix (group_power.lean)

abhimanyupallavisudhir committed:
fix(exponential.lean)

abhimanyupallavisudhir committed:
fix(data/complex/exponential.lean)

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -397,6 +423,70 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+quotient.sound' ⟨-1, by dsimp only; rw [neg_one_gsmul, add_zero]⟩
+
+theorem of_cos_eq {θ ψ : ℝ} (Hcos : cos θ = cos ψ) : (θ : angle) = ψ ∨ (θ : angle) = -ψ :=
+begin
```
`cos_eq_iff_eq_or_eq_neg` and iff and only iff

ChrisHughes24 commented on analysis/exponential.lean:
```diff
@@ -397,6 +423,70 @@ lemma exists_sin_eq {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : ∃ y, -(π
+end
+
+theorem of_sin_eq {θ ψ : ℝ} (Hsin : sin θ = sin ψ) : (θ : angle) = ψ ∨ (θ : angle) + ψ = π :=
+begin
```
same as above

abhimanyupallavisudhir replied:
Done.

abhimanyupallavisudhir committed:
fix(analysis/exponential): cos_eq_iff_eq_or_eq_neg

abhimanyupallavisudhir committed:
Merge remote-tracking branch 'origin/master' into xena
